### ISSUE TITLE
Support JSON encoder/decoder for session data

### ISSUE THREAD

**Issue by rskumar** [alexedwards/scs#68]
Currently `scs` supports data encoding/decoding through only `gob`. 
`gob` encoded data can only be used by golang based application.
Letting app developer decide the encoder/decoder of their choice (interface contract), would be great feature. Or, at-least, JSON encoder/decoder support is needed.

**Comment by alexedwards** [comment]
In theory it's possible to encode/decode the session data to JSON, but the downside of using JSON is that there's no guarantee that the decoded data will have the same type as the original data.

For example, if you encode a `[]byte` slice with `encoding/json` it will be stored as a base64-encoded string in the JSON result. On decoding, that is then decoded to a `string` rather than the original `[]byte`. The same thing is true of `int` type and some other types like `float32`.

That doesn't mean that JSON encoding can't work, just that it results in some behaviour which may be confusing for clients of the `scs` package. Which is why I opted to use `gob` only in the current version. 

Your idea of letting the developer set their own encoder/decoder via an interface might be the better option. It sounds like it would be more flexible, but also would also force people to be more aware of exactly what their choice of encoder is doing. 

I will think on this for a few days. 

On the flip side, could you describe your use-case in a bit more detail? There might be an alternative solution rather than changing the way `scs` works.

**Comment by rskumar** [comment]
> On the flip side, could you describe your use-case in a bit more detail? There might be an alternative solution rather than changing the way scs works.

Sure @alexedwards . There are few scenarios where json would help me.
First, I have multi-tenant app. A user, with same email can be part of one or more tenant. He can login into multiple account and mark one as current active. I am planning to keep user's multiple login info in single session in list with one as marked.  I can do this with gob or json.

The admin of one tenant can invaliate that user's active session (but that does not kill other login data) . This admin app is in python. It can search for user keys in redis, read data, but cannot change as its int `gob`. If its JSON, it would directly modify and store back in redis.

**Comment by alexedwards** [comment]
What I'm thinking is to create a `Serializer` interface like this:

```go
type Serializer interface {
	Encode(deadline time.Time, values map[string]interface{}) ([]byte, error)
	Decode([]byte) (deadline time.Time, values map[string]interface{}, err error)
}
```

A user can then define their own custom `Serializer`. For instance, to serialize with JSON you could define something like this:

```go
type jsonSerializer struct{}

func (_ jsonSerializer) Encode(deadline time.Time, values map[string]interface{}) ([]byte, error) {
	return json.Marshal(&struct {
		Deadline int64                  `json:"deadline"`
		Values   map[string]interface{} `json:"values"`
	}{
		Deadline: deadline.UnixNano(),
		Values:   values,
	})
}

func (_ jsonSerializer) Decode(b []byte) (time.Time, map[string]interface{}, error) {
	aux := struct {
		Deadline int64                  `json:"deadline"`
		Values   map[string]interface{} `json:"values"`
	}{}

	dec := json.NewDecoder(bytes.NewReader(b))
	dec.UseNumber()
	err := dec.Decode(&aux)
	if err != nil {
		return time.Time{}, nil, err
	}
	return time.Unix(0, aux.Deadline), aux.Values, nil
}
```

This can then be set as the serializer when initializing the session manager. For example:

```go
sessionManager := scs.New()
sessionManager.Serializer = jsonSerializer{}
sessionManager.Lifetime = time.hour
```

Would that work for you? 

I'm also not 100% happy about the name `Serializer`. What do you think? Or do you have any alternative suggestions?

**Comment by rskumar** [comment]
@alexedwards thanks. Yes that would work for me.

Meanwhile, I was also looking a way around, considering it a responsibility of `store` to encode/decode data as they like,. But that means change in current design, breaking backward compatibility, so cannot go with :( . Maybe in some next major version.

**Comment by alexedwards** [comment]
This is fixed by commit 4cd374bf181ed45ace1a3f9af3a04cc26ba70243. You can now define your own `Codec` like so:

```
type jsonCodec struct{}

func (jsonCodec) Encode(deadline time.Time, values map[string]interface{}) ([]byte, error) {
	return json.Marshal(&struct {
		Deadline int64                  `json:"deadline"`
		Values   map[string]interface{} `json:"values"`
	}{
		Deadline: deadline.UnixNano(),
		Values:   values,
	})
}

func (jsonCodec) Decode(b []byte) (time.Time, map[string]interface{}, error) {
	aux := struct {
		Deadline int64                  `json:"deadline"`
		Values   map[string]interface{} `json:"values"`
	}{}

	dec := json.NewDecoder(bytes.NewReader(b))
	dec.UseNumber()
	err := dec.Decode(&aux)
	if err != nil {
		return time.Time{}, nil, err
	}
	return time.Unix(0, aux.Deadline), aux.Values, nil
}
```

And specify this when setting up the session manager, like so:

```
sessionManager := scs.New()
sessionManager.Serializer = jsonCodec{}
sessionManager.Lifetime = time.hour
```
